### PR TITLE
🚀 Vehicle Improvements

### PR DESCRIPTION
**Pull request by AyhamAl-Ali**
### Description

- :heavy_check_mark: Added more vehicle events (Collision, Move)
- :heavy_check_mark: Added `vehicle` expression
- :heavy_check_mark: Added vehicle collision cause expression (see note below)
- :heavy_check_mark: Fixed event values of vehicles
- :heavy_check_mark: Changed Vehicle PropertyExpression to VehicleOf
- :heavy_check_mark: Changed Vehicle Effect name to Ride / Mount (Vehicle here doesn't make sense and also to not conflict)

**NOTE:** Vehicle collision cause currently returns an object type which probably makes it lose some of its information as a block or an entity as I have tried to collide with a zombie named "Hi" and executing `name of collision cause` only returned "zombie"

---
**Target Minecraft Versions:** Any
**Requirements:** None
**Related Issues:** #4328 #1337

### CODE REVIEW

File: src/main/java/ch/njol/skript/events/EvtVehicleCollision.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+public class EvtVehicleCollision extends SkriptEvent {
+
+	static {
+		Class<? extends Event>[] events = CollectionUtils.array(VehicleBlockCollisionEvent.class, VehicleEntityCollisionEvent.class);
+
+		Skript.registerEvent("Vehicle Collision", EvtVehicleCollision.class, events, "vehicle [(1¦block|2¦entity)] collision", "(1¦block|2¦entity|) collision of [a] vehicle")
+			.description("Called when a vehicle collides with an entity or a block.")
+			.examples("on vehicle move:",
+				"\t\tpush vehicle upwards with force 1.5")
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/events/EvtVehicleCollision.java:39]: Self note: use only `\t`

File: src/main/java/ch/njol/skript/expressions/ExprVehicle.java [hunk 2]
```diff
@@ -18,126 +18,61 @@
...
-		assert false;
-		return e.getVehicle();
+		Skript.registerExpression(ExprVehicle.class, Vehicle.class, ExpressionType.SIMPLE, "[the] vehicle");
 	}
 	
 	@Override
-	public Class<? extends Entity> getReturnType() {
-		return Entity.class;
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parser) {
+		System.out.println(getParser().getCurrentEvents()[0].toString());
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprVehicle.java:50]: Self note: debug statement

File: src/main/java/ch/njol/skript/expressions/ExprVehicle.java [hunk 3]
```diff
@@ -18,126 +18,61 @@
...
+		if (!VehicleEvent.class.isAssignableFrom(getParser().getCurrentEvents()[0])) {
+			Skript.error("Cannot use 'vehicle' outside of a vehicle related events.");
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprVehicle.java:52]: Self note: remove period

File: src/main/java/ch/njol/skript/expressions/ExprVehicleCollisionCause.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	"\t\tkill event-entity",
+	"",
+	"on vehicle block collision:",
+	"\tif collision cause is a block:",
+	"\t\tdamage event-entity by 1 heart"})
+@Since("INSERT VERSION")
+public class ExprVehicleCollisionCause extends SimpleExpression<Object> {
+
+	static {
+		Skript.registerExpression(ExprVehicleCollisionCause.class, Object.class, ExpressionType.SIMPLE, "[the] [vehicle] collision cause");
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprVehicleCollisionCause.java:49]: Self note: Would be better to make `vehicle` part required

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 5]
```diff
@@ -106,11 +106,7 @@
 import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 import org.bukkit.event.player.PlayerToggleFlightEvent;
 import org.bukkit.event.server.ServerCommandEvent;
-import org.bukkit.event.vehicle.VehicleDamageEvent;
-import org.bukkit.event.vehicle.VehicleDestroyEvent;
-import org.bukkit.event.vehicle.VehicleEnterEvent;
-import org.bukkit.event.vehicle.VehicleEvent;
-import org.bukkit.event.vehicle.VehicleExitEvent;
+import org.bukkit.event.vehicle.*;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:109]: We can't be doing wild card imports please

File: src/main/java/ch/njol/skript/events/SimpleEvents.java [hunk 6]
```diff
@@ -85,11 +85,7 @@
 import org.bukkit.event.player.PlayerToggleSneakEvent;
 import org.bukkit.event.player.PlayerToggleSprintEvent;
 import org.bukkit.event.server.ServerListPingEvent;
-import org.bukkit.event.vehicle.VehicleCreateEvent;
-import org.bukkit.event.vehicle.VehicleDamageEvent;
-import org.bukkit.event.vehicle.VehicleDestroyEvent;
-import org.bukkit.event.vehicle.VehicleEnterEvent;
-import org.bukkit.event.vehicle.VehicleExitEvent;
+import org.bukkit.event.vehicle.*;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/events/SimpleEvents.java:88]: ^ wild card

File: src/main/java/ch/njol/skript/expressions/ExprVehicleOf.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+				}
+				return p.getVehicle();
+			}
+		});
+	}
+	
+	@Override
+	@Nullable
+	public Entity convert(final Entity e) {
+		assert false;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprVehicleOf.java:85]: Is this assert false required?

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

**Comment by TheLimeGlass** [comment]
Say you have a list of multiple entities and they're going to be set onto 1 vehicle, wouldn't it be a better implementation to have all the entities stack onto eachother riding the set vehicle 
image

**Comment by TPGamesNL** [comment]
You should create the vehicle entity data instead of the vehicle expression (e.g. just like player or any other entity)

**Comment by sovdeeth** [comment]
Closing due to inactivity.

**Pull request closed without merging**